File: benchmarks/multilang-wasm/numeric-polybench-panel/polybench.dart
// numeric-polybench-panel Dart WasmGC kernel
// Mirrors benchmarks/base/numeric-polybench-panel/workload.js & WAT/reference.c:
// 4 kernels: gemm, cholesky, stencil5, jacobi2d operating on double precision (f64).

import 'dart:js_interop';
import 'dart:math' as math;

@JSExport()
class PolybenchKernels {
  @JSExport('gemm')
  void gemm(
    JSFloat64Array aJs,
    JSFloat64Array bJs,
    JSFloat64Array cJs,
    int ni,
    int nj,
    int nk,
    double alpha,
    double beta,
  ) {
    final a = aJs.toDart;
    final b = bJs.toDart;
    final c = cJs.toDart;
    for (int i = 0; i < ni; ++i) {
      for (int j = 0; j < nj; ++j) c[i * nj + j] *= beta;
      for (int k = 0; k < nk; ++k) {
        final aik = a[i * nk + k];
        for (int j = 0; j < nj; ++j) {
          c[i * nj + j] += alpha * aik * b[k * nj + j];
        }
      }
    }
  }

  @JSExport('cholesky')
  int cholesky(JSFloat64Array aJs, int n) {
    final a = aJs.toDart;
    for (int i = 0; i < n; ++i) {
      for (int j = 0; j < i; ++j) {
        for (int k = 0; k < j; ++k) {
          a[i * n + j] -= a[i * n + k] * a[j * n + k];
        }
        a[i * n + j] /= a[j * n + j];
      }
      for (int k = 0; k < i; ++k) {
        a[i * n + i] -= a[i * n + k] * a[i * n + k];
      }
      final aii = a[i * n + i];
      if (!(aii > 0.0)) return 0;
      a[i * n + i] = math.sqrt(aii);
      for (int j = i + 1; j < n; ++j) {
        a[i * n + j] = 0.0;
      }
    }
    return 1;
  }

  @JSExport('stencil5')
  void stencil5(JSFloat64Array aJs, JSFloat64Array outJs, int n) {
    final a = aJs.toDart;
    final out = outJs.toDart;
    for (int i = 1; i < n - 1; ++i) {
      for (int j = 1; j < n - 1; ++j) {
        int p = i * n + j;
        out[p] = 0.2 * (a[p] + a[p - 1] + a[p + 1] + a[p - n] + a[p + n]);
      }
    }
  }

  @JSExport('jacobi2d')
  void jacobi2d(JSFloat64Array aJs, JSFloat64Array bJs, int n, int timesteps) {
    final a = aJs.toDart;
    final b = bJs.toDart;
    for (int t = 0; t < timesteps; ++t) {
      for (int i = 1; i < n - 1; ++i) {
        for (int j = 1; j < n - 1; ++j) {
          int p = i * n + j;
          b[p] = 0.2 * (a[p] + a[p - 1] + a[p + 1] + a[p - n] + a[p + n]);
        }
      }
      for (int i = 1; i < n - 1; ++i) {
        for (int j = 1; j < n - 1; ++j) {
          int p = i * n + j;
          a[p] = 0.2 * (b[p] + b[p - 1] + b[p + 1] + b[p - n] + b[p + n]);
        }
      }
    }
  }
}

void main() {
  dartKernels = createJSInteropWrapper(PolybenchKernels());
}

@JS('dartKernels')
external set dartKernels(JSObject value);

File: benchmarks/multilang-wasm/numeric-polybench-panel/polybench.ts
// polybench.ts — AssemblyScript multilang kernel for numeric.polybench-panel.v1.
//
// Mirrors polybench.c and benchmarks/base/numeric-polybench-panel/workload.js:
// four f64 kernels — gemm, cholesky, stencil5 and jacobi2d — in the same loop
// order with the same term ordering, so every rounding step matches.
//
// AssemblyScript emits unfused f64 mul and add, matching the C build's
// -ffp-contract=off and the JS oracle's scalar accumulation. Pointers are raw
// linear-memory byte offsets; no allocation, no runtime imports.

export function gemm(
  a: usize,
  b: usize,
  c: usize,
  ni: i32,
  nj: i32,
  nk: i32,
  alpha: f64,
  beta: f64,
): void {
  for (let i = 0; i < ni; ++i) {
    for (let j = 0; j < nj; ++j) {
      const p: usize = c + (<usize> (i * nj + j)) * 8;
      store<f64>(p, load<f64>(p) * beta);
    }
    for (let k = 0; k < nk; ++k) {
      const av: f64 = load<f64>(a + (<usize> (i * nk + k)) * 8);
      for (let j = 0; j < nj; ++j) {
        const p: usize = c + (<usize> (i * nj + j)) * 8;
        store<f64>(p, load<f64>(p) + alpha * av * load<f64>(b + (<usize> (k * nj + j)) * 8));
      }
    }
  }
}

export function cholesky(a: usize, n: i32): i32 {
  for (let i = 0; i < n; ++i) {
    for (let j = 0; j < i; ++j) {
      const pij: usize = a + (<usize> (i * n + j)) * 8;
      for (let k = 0; k < j; ++k) {
        store<f64>(
          pij,
          load<f64>(pij) -
            load<f64>(a + (<usize> (i * n + k)) * 8) * load<f64>(a + (<usize> (j * n + k)) * 8),
        );
      }
      store<f64>(pij, load<f64>(pij) / load<f64>(a + (<usize> (j * n + j)) * 8));
    }
    const pii: usize = a + (<usize> (i * n + i)) * 8;
    for (let k = 0; k < i; ++k) {
      const v: f64 = load<f64>(a + (<usize> (i * n + k)) * 8);
      store<f64>(pii, load<f64>(pii) - v * v);
    }
    if (!(load<f64>(pii) > 0.0)) return 0;
    store<f64>(pii, Math.sqrt(load<f64>(pii)));
    for (let j = i + 1; j < n; ++j) {
      store<f64>(a + (<usize> (i * n + j)) * 8, 0.0);
    }
  }
  return 1;
}

export function stencil5(a: usize, out: usize, n: i32): void {
  for (let i = 1; i < n - 1; ++i) {
    for (let j = 1; j < n - 1; ++j) {
      const p: i32 = i * n + j;
      const base: usize = a + (<usize> p) * 8;
      const sum: f64 = load<f64>(base) + load<f64>(base - 8) + load<f64>(base + 8) +
        load<f64>(a + (<usize> (p - n)) * 8) + load<f64>(a + (<usize> (p + n)) * 8);
      store<f64>(out + (<usize> p) * 8, 0.2 * sum);
    }
  }
}

export function jacobi2d(a: usize, b: usize, n: i32, timesteps: i32): void {
  for (let t = 0; t < timesteps; ++t) {
    stencil5(a, b, n);
    stencil5(b, a, n);
  }
}
